Mathlib/Computability/Language.lean

```diff
@@ -176,24 +177,25 @@ theorem map_map (g : β → γ) (f : α → β) (l : Language α) : map g (map f
 #align language.map_map Language.map_map
 
 theorem star_def_nonempty (l : Language α) :
-    l.star = { x | ∃ S : List (List α), x = S.join ∧ ∀ y ∈ S, y ∈ l ∧ y ≠ [] } :=
-  by
+    l.star = { x | ∃ S : List (List α), x = S.join ∧ ∀ y ∈ S, y ∈ l ∧ y ≠ [] } := by
   ext x
   constructor
   · rintro ⟨S, rfl, h⟩
-    refine' ⟨S.filter fun l => ¬List.isEmpty l, by simp, fun y hy => _⟩
-    rw [mem_filter, empty_iff_eq_nil] at hy
-    exact ⟨h y hy.1, hy.2⟩
+    refine' ⟨S.filter λ l ↦ ¬List.isEmpty l, by simp, λ y hy ↦ _⟩
+    simp [mem_filter, List.isEmpty_iff_eq_nil] at hy
+    let ⟨hyl, hyr⟩ := hy
+    apply And.intro (h y hyl)
+    cases y <;> simp only [ne_eq, not_true, not_false_iff]
+    contradiction
```

Comment:
Why does the one-liner in the old code no longer work?

Reply:
The goal `y ≠ []` cannot be derived from `hy.2 : isEmpty y = false`.

Reply:
Please record that observation in a porting note. We might want to add a lemma that links those two facts, and golf this proof afterwards.